File: apps/app/src/lib/user-display.ts
export function formatWalletAddress(address?: string | null): string {
  const value = address?.trim();
  if (!value) return "";
  if (value.length <= 8) return value;
  return `${value.slice(0, 3)}...${value.slice(-3)}`;
}

export function getUserDisplayName(input: {
  fullName?: string | null;
  email?: string | null;
  walletAddress?: string | null;
  fallback?: string;
}): string {
  const fullName = input.fullName?.trim();
  if (fullName) return fullName;

  const emailLocalPart = input.email?.split("@")[0]?.trim();
  if (emailLocalPart) return emailLocalPart;

  const formattedWalletAddress = formatWalletAddress(input.walletAddress);
  if (formattedWalletAddress) return formattedWalletAddress;

  return input.fallback ?? "User";
}

File: apps/ios/Packages/AggrCore/Sources/AggrCore/User/UserDisplay.swift
import Foundation

/// Port of `apps/app/src/lib/user-display.ts`.
public enum UserDisplay {
  public static func formatWalletAddress(_ address: String?) -> String {
    guard let value = address?.trimmingCharacters(in: .whitespacesAndNewlines), !value.isEmpty else { return "" }
    if value.count <= 8 { return value }
    return "\(value.prefix(3))...\(value.suffix(3))"
  }

  public static func displayName(fullName: String?, email: String?, walletAddress: String?, fallback: String = "User") -> String {
    if let fullName = fullName?.trimmingCharacters(in: .whitespacesAndNewlines), !fullName.isEmpty { return fullName }
    if let local = email?.split(separator: "@").first.map({ String($0).trimmingCharacters(in: .whitespaces) }), !local.isEmpty { return local }
    let wallet = formatWalletAddress(walletAddress)
    if !wallet.isEmpty { return wallet }
    return fallback
  }
}
